File: armemu/gen_dp.pl
#!/usr/bin/perl -w

@iname = ("and", "eor", "sub", "rsb", "add", "adc", "sbc", "rsc",
          "tst", "teq", "cmp", "cmn", "orr", "mov", "bic", "mvn");

%islogic = ("and", 1, "eor", 1, "sub", 0, "rsb", 0,
            "add", 0, "adc", 0, "sbc", 0, "rsc", 0,
            "tst", 1, "teq", 1, "cmp", 0, "cmn", 0,
            "orr", 1, "mov", 1, "bic", 1, "mvn", 1);

@sname = ("lsl", "lsr", "asr", "ror");

print ".text\n";

foreach $inst (@iname)
{
  foreach $setflags ('s', 'ns')
  {
    foreach $shifttype (@sname)
    {
      foreach $regshift ('reg', 'imm')
      {
        $funcname = "${inst}_${setflags}_${shifttype}_${regshift}";
        print ".p2align 4,,7\n";
        print ".globl $funcname\n";
        print ".type $funcname,\@function\n";
        print "$funcname:\n";
        
        outputShift();
        
        print "\tmovl 4(\%esp),\%eax\n";
        
        $invc = 0;
        $needtest = 0;
        
        if ($inst eq 'and') {
          print "\tandl \%edx,\%eax\n";
        }
        elsif ($inst eq 'eor') {
          print "\txorl \%edx,\%eax\n";
        }
        elsif ($inst eq 'sub') {
          print "\tsubl \%edx,\%eax\n";
          $invc = 1;
        }
        elsif ($inst eq 'rsb') {
          print "\txchgl \%edx,\%eax\n";
          print "\tsubl \%edx,\%eax\n";
        }
        elsif ($inst eq 'add') {
          print "\taddl \%edx,\%eax\n";
        }
        elsif ($inst eq 'adc') {
          print "\tmovl 16(\%esp),\%ecx\n";
          print "\tbt \$29,(\%ecx)\n";
          print "\tadcl \%edx,\%eax\n";
        }
        elsif ($inst eq 'sbc') {
          print "\tmovl 16(\%esp),\%ecx\n";
          print "\tbt \$29,(\%ecx)\n";
          print "\tcmc\n";
          print "\tsbbl \%edx,\%eax\n";
          $invc = 1;
        }
        elsif ($inst eq 'rsc') {
          print "\tmovl 16(\%esp),\%ecx\n";
          print "\txchgl \%edx,\%eax\n";
          print "\tbt \$29,(\%ecx)\n";
          print "\tcmc\n";
          print "\tsbbl \%edx,\%eax\n";
          $invc = 1;
        }
        elsif ($inst eq 'tst') {
          print "\ttestl \%edx,\%eax\n";
        }
        elsif ($inst eq 'teq') {
          print "\tmovl \%eax,\%ecx\n";
          print "\txorl \%edx,\%ecx\n";
        }
        elsif ($inst eq 'cmp') {
          print "\tcmpl \%edx,\%eax\n";
          $invc = 1;
        }
        elsif ($inst eq 'cmn') {
          print "\tmovl \%eax,\%ecx\n";
          print "\taddl \%edx,\%ecx\n";
        }
        elsif ($inst eq 'orr') {
          print "\torl \%edx,\%eax\n";
        }
        elsif ($inst eq 'mov') {
          print "\tmovl \%edx,\%eax\n";
          $needtest = 1;
        }
        elsif ($inst eq 'bic') {
          print "\tnotl \%edx\n";
          print "\tandl \%edx,\%eax\n";
        }
        elsif ($inst eq 'mvn') {
          print "\tnotl \%edx\n";
          print "\tmovl \%edx,\%eax\n";
          $needtest = 1;
        }
        
        if ($setflags eq 's')
        {
          if ($islogic{$inst})
          {
            if ($needtest) {
              print "\ttestl \%eax,\%eax\n";
            }
            print "\tsete \%dl\n";
            print "\tsets \%dh\n";
            print "\tshl \$1,\%dh\n";
            print "\tor \%dh,\%dl\n";
            print "\tshl \$30,\%edx\n";
            print "\tmovl 16(\%esp),\%ecx\n";
            print "\tandl \$0x3fffffff,(\%ecx)\n";
            print "\torl \%edx,(\%ecx)\n";
          }
          else  # arithmetic instruction
          {
            print "\tseto \%dl\n";     # V
            if ($invc) {
              print "\tsetnc \%dh\n";  # !C
            }
            else {
              print "\tsetc \%dh\n";   # C
            }
            print "\tsete \%cl\n";     # Z
            print "\tsets \%ch\n";     # N
            print "\tshlb \$1,\%dh\n";
            print "\tshlb \$2,\%cl\n";
            print "\tshlb \$3,\%ch\n";
            print "\torb \%dh,\%dl\n";
            print "\torb \%cl,\%dl\n";
            print "\torb \%ch,\%dl\n";
            print "\tshll \$28,\%edx\n";
            print "\tmovl 16(\%esp),\%ecx\n";
            print "\tandl \$0x0fffffff,(\%ecx)\n";
            print "\torl \%edx,(\%ecx)\n";
          }
        }
        print "\tret\n\n";
      }
    }
  }
}

sub outputShift
{
  my ($flags) = $setflags eq 's' && $islogic{$inst};

  my $label = sub {
    return ($funcname . '_' . $_[0] . ":\n");
  };

  my $ref = sub {
    return ($funcname . '_' . $_[0] . "\n");
  };

  if ($regshift eq 'reg')
  {
    if ($shifttype eq 'lsl')
    {
      print "\tmovzbl 12(\%esp),\%ecx\n";
      print "\tmovl 8(\%esp),\%edx\n";
      print "\tcmp \$31,\%cl\n";
      print "\tjg " . &$ref(1);
      print "\tshll \%cl,\%edx\n";
      if ($flags)
      {
        print "\tsetc \%cl\n";
        print "\tshll \$29,\%ecx\n";
        print "\tmovl 16(\%esp),\%eax\n";
        print "\tbtr \$29,(\%eax)\n";
        print "\torl \%ecx,(\%eax)\n";
      }
      print "\tjmp " . &$ref(2);
      print &$label(1);
      print "\tcmp \$32,\%cl\n";
      print "\tje " . &$ref(3);
      # greater than 32-bit shift
      if ($flags)
      {
        print "\tbtr \$29,(\%eax)\n"; # clear carry
      }
      print "\txorl \%edx,\%edx\n";   # clear edx
      print "\tjmp " . &$ref(2);
      print &$label(3);
      if ($flags)
      {
        print "\tbt \$31,\%edx\n";
        print "\tsetc \%cl\n";
        print "\tshll \$29,\%ecx\n";
        print "\tbtr \$29,(\%eax)\n";
        print "\torl \%ecx,(\%eax)\n";
      }
      print "\txorl \%edx,\%edx\n";
      print &$label(2);
    }
    elsif ($shifttype eq 'lsr')
    {
      print "\tmovzbl 12(\%esp),\%ecx\n";
      print "\tmovl 8(\%esp),\%edx\n";
      print "\tcmp \$31,\%cl\n";
      print "\tjg " . &$ref(1);
      print "\tshrl \%cl,\%edx\n";
      if ($flags)
      {
        print "\tsetc \%cl\n";
        print "\tshll \$29,\%ecx\n";
        print "\tmovl 16(\%esp),\%eax\n";
        print "\tbtr \$29,(\%eax)\n";
        print "\torl \%ecx,(\%eax)\n";
      }
      print "\tjmp " . &$ref(2);
      print &$label(1);
      print "\tcmp \$32,\%cl\n";
      print "\tje " . &$ref(3);
      # greater than 32-bit shift
      if ($flags)
      {
        print "\tbtr \$29,(\%eax)\n"; # clear carry
      }
      print "\txorl \%edx,\%edx\n";   # clear edx
      print "\tjmp " . &$ref(2);
      print &$label(3);
      if ($flags)
      {
        print "\tbt \$0,\%edx\n";
        print "\tsetc \%cl\n";
        print "\tshll \$29,\%ecx\n";
        print "\tbtr \$29,(\%eax)\n";
        print "\torl \%ecx,(\%eax)\n";
      }
      print "\txorl \%edx,\%edx\n";
      print &$label(2);    
    }
    elsif ($shifttype eq 'asr')
    {
      print "\tmovzbl 12(\%esp),\%ecx\n";
      print "\tmovl 8(\%esp),\%edx\n";
      print "\tcmp \$31,\%cl\n";
      print "\tjg " . &$ref(1);
      print "\tsarl \%cl,\%edx\n";
      if ($flags)
      {
        print "\tsetc \%cl\n";
        print "\tshll \$29,\%ecx\n";
        print "\tmovl 16(\%esp),\%eax\n";
        print "\tbtr \$29,(\%eax)\n";
        print "\torl \%ecx,(\%eax)\n";
      }
      print "\tjmp " . &$ref(2);
      print &$label(1);
      # greater than 31-bit shift
      print "\tbt \$31,\%edx\n";
      print "\txorl \%edx,\%edx\n";  # clear edx
      print "\tsbbl \$0,\%edx\n";
      if ($flags)
      {
        print "\tmovl \%edx,\%eax\n";
        print "\tandl \$0x20000000,\%eax\n";
        print "\tbtr \$29,(\%eax)\n";
        print "\torl \%eax,(\%eax)\n";
      }
      print &$label(2);
    }
    elsif ($shifttype eq 'ror')
    {
      print "\tmovzbl 12(\%esp),\%ecx\n";
      print "\tmovl 8(\%esp),\%edx\n";
      # register-valued ROR of zero doesn't seem to affect C
      if ($flags)
      {
        print "\tcmp \$0,\%cl\n";
        print "\tje " . &$ref(1);
      }
      print "\trorl \%cl,\%edx\n";
      if ($flags)
      {
        print "\tsetc \%cl\n";
        print "\tshll \$29,\%ecx\n";
        print "\tmovl 16(\%esp),\%eax\n";
        print "\tbtr \$29,(\%eax)\n";
        print "\torl \%ecx,(\%eax)\n";
        print &$label(1);
      }
    }
  }
  else  # immediate shift
  {
    if ($shifttype eq 'lsl')
    {
      print "\tmovl 12(\%esp),\%ecx\n";
      print "\tmovl 8(\%esp),\%edx\n";
      print "\tshll \%cl,\%edx\n";
      if ($flags)
      {
        print "\tsetc \%cl\n";
        print "\tshll \$29,\%ecx\n";
        print "\tmovl 16(\%esp),\%eax\n";
        print "\tbtr \$29,(\%eax)\n";
        print "\torl \%ecx,(\%eax)\n";
      }
    }
    elsif ($shifttype eq 'lsr')
    {
      if ($flags)
      {
        print "\tmovl 12(\%esp),\%ecx\n";
        print "\tcmpl \$0,\%ecx\n";
        print "\tje " . &$ref(1);
        print "\tmovl 8(\%esp),\%edx\n";
        print "\tshrl \%cl,\%edx\n";
        print "\tsetc \%cl\n";
        print "\tshll \$29,\%ecx\n";
        print "\tmovl 16(\%esp),\%eax\n";
        print "\tbtr \$29,(\%eax)\n";
        print "\torl \%ecx,(\%eax)\n";
        print "\tjmp " . &$ref(2);
        print &$label(1);
        print "\tbt \$31,\%edx\n";
        print "\tsetc \%cl\n";
        print "\tshll \$29,\%ecx\n";
        print "\tbtr \$29,(\%eax)\n";
        print "\torl \%ecx,(\%eax)\n";
        print &$label(2);
      }
      else
      {
        print "\tmovl 12(\%esp),\%ecx\n";
        print "\tcmpl \$0,\%ecx\n";
        print "\tje " . &$ref(1);
        print "\tmovl 8(\%esp),\%edx\n";
        print "\tshrl \%cl,\%edx\n";
        print "\tjmp " . &$ref(2);
        print &$label(1);
        print "\tmovl 4(\%esp),\%eax\n";
        print "\tret\n";
        print &$label(2);
      }
    }
    elsif ($shifttype eq 'asr')
    {
      if ($flags)
      {
        print "\tmovl 12(\%esp),\%ecx\n";
        print "\tcmpl \$0,\%ecx\n";
        print "\tje " . &$ref(1);
        print "\tmovl 8(\%esp),\%edx\n";
        print "\tsarl \%cl,\%edx\n";
        print "\tsetc \%cl\n";
        print "\tshll \$29,\%ecx\n";
        print "\tmovl 16(\%esp),\%eax\n";
        print "\tbtr \$29,(\%eax)\n";
        print "\torl \%ecx,(\%eax)\n";
        print "\tjmp " . &$ref(2);
        print &$label(1);
        print "\tbt \$31,\%edx\n";
        print "\tsetc \%cl\n";
        print "\tsarl \$31,\%edx\n";
        print "\tshll \$29,\%ecx\n";
        print "\tbtr \$29,(\%eax)\n";
        print "\torl \%ecx,(\%eax)\n";
        print &$label(2);
      }
      else
      {
        print "\tmovl 12(\%esp),\%ecx\n";
        print "\tcmpl \$0,\%ecx\n";
        print "\tje " . &$ref(1);
        print "\tmovl 8(\%esp),\%edx\n";
        print "\tsarl \%cl,\%edx\n";
        print "\tjmp " . &$ref(2);
        print &$label(1);
        print "\tmovl 4(\%esp),\%eax\n";
        print "\tret\n";
        print &$label(2);
      }
    }
    elsif ($shifttype eq 'ror')
    {
      print "\tmovl 12(\%esp),\%ecx\n";
      print "\tmovl 8(\%esp),\%edx\n";
      print "\tmovl 16(\%esp),\%eax\n";
      print "\tcmpl \$0,\%ecx\n";
      print "\tje " . &$ref(1);
      print "\trorl \%cl,\%edx\n";
      print "\tjmp " . &$ref(2);
      print &$label(1);
      print "\tbt \$29,(\%eax)\n";
      print "\trcrl \$1,\%edx\n";
      print &$label(2);
      # carry valid for either ROR or RRX operation
      if ($flags)
      {
        print "\tsetc \%cl\n";
        print "\tshll \$29,\%ecx\n";
        print "\tbtr \$29,(\%eax)\n";
        print "\torl \%ecx,(\%eax)\n";
      }
    }
  }
}

File: armemu/gen_dp.c
#include <stdio.h>

#include "defs.h"
#include "machine.h"
#include "memory.h"
#include "registers.h"
#include "processor.h"
#include "execute.h"
#include "fakesys.h"
#include "decode.h"

// generate nice assembly language leaf functions to deal with all that
// data-processing flag crap
void gen_dp(FILE* stream)
{
  int op;
  const char* iname[] =
    {"and", "eor", "sub", "rsb", "add", "adc", "sbc", "rsc",
     "tst", "teq", "cmp", "cmn", "orr", "mov", "bic", "mvn"};
  const char* sname[] =
    {"lsl", "lsr", "asr", "ror"};
  const uint5 logic = 0xf303;
  
  fprintf(stream, ".text\n");
  
  for (op=0; op<16; op++)
  {
    int setflags, islogic = logic & (1<<op);
    for (setflags=0; setflags<2; setflags++)
    {
      int shifttype;
      for (shifttype=0; shifttype<4; shifttype++)
      {
        int regshift;
        for (regshift=0; regshift<2; regshift++)
        {
          int invc = 0, needtest = 0;
#define fnid iname[op], setflags?"s":"ns", sname[shifttype], \
             regshift?"reg":"imm"
#define ref(N) "%s_%s_%s_%s_" #N "\n\t", fnid
#define label(N) fprintf(stream, "%s_%s_%s_%s_" #N ":\n\t", fnid);
          fprintf(stream, ".p2align 4,,7\n\t");
          fprintf(stream, ".globl %s_%s_%s_%s\n\t", fnid);
          fprintf(stream, ".type %s_%s_%s_%s,@function\n", fnid);
          fprintf(stream, "%s_%s_%s_%s:\n\t", fnid);

// flags go:
//   bit 31: n
//   bit 30: z
//   bit 29: c
//   bit 28: v

          if (regshift)
          {
            switch (shifttype)
            {
              case 0:  // lsl
              fprintf(stream, "movzbl 12(%%esp),%%ecx\n\t");
              fprintf(stream, "movl 8(%%esp),%%edx\n\t");
              fprintf(stream, "cmp $31,%%cl\n\t");
              fprintf(stream, "jg "ref(1));
              fprintf(stream, "shll %%cl,%%edx\n\t");
              if (setflags && islogic)
              {
                fprintf(stream, "setc %%cl\n\t");
                fprintf(stream, "shll $29,%%ecx\n\t");
                fprintf(stream, "movl 16(%%esp),%%eax\n\t");
                fprintf(stream, "btr $29,(%%eax)\n\t");
                fprintf(stream, "orl %%ecx,(%%eax)\n\t");
              }
              fprintf(stream, "jmp "ref(2));
              label(1);
              fprintf(stream, "cmp $32,%%cl\n\t");
              fprintf(stream, "je "ref(3));
              // greater than 32-bit shift
              if (setflags && islogic)
              {
                fprintf(stream, "btr $29,(%%eax)\n\t"); // clear carry
              }
              fprintf(stream, "xorl %%edx,%%edx\n\t");  // clear edx
              fprintf(stream, "jmp "ref(2));
              label(3);
              if (setflags && islogic)
              {
                fprintf(stream, "bt $31,%%edx\n\t");
                fprintf(stream, "setc %%cl\n\t");
                fprintf(stream, "shll $29,%%ecx\n\t");
                fprintf(stream, "btr $29,(%%eax)\n\t");
                fprintf(stream, "orl %%ecx,(%%eax)\n\t");
              }
              fprintf(stream, "xorl %%edx,%%edx\n\t");
              label(2);
              break;
              
              case 1:  // lsr
              fprintf(stream, "movzbl 12(%%esp),%%ecx\n\t");
              fprintf(stream, "movl 8(%%esp),%%edx\n\t");
              fprintf(stream, "cmp $31,%%cl\n\t");
              fprintf(stream, "jg "ref(1));
              fprintf(stream, "shrl %%cl,%%edx\n\t");
              if (setflags && islogic)
              {
                fprintf(stream, "setc %%cl\n\t");
                fprintf(stream, "shll $29,%%ecx\n\t");
                fprintf(stream, "movl 16(%%esp),%%eax\n\t");
                fprintf(stream, "btr $29,(%%eax)\n\t");
                fprintf(stream, "orl %%ecx,(%%eax)\n\t");
              }
              fprintf(stream, "jmp "ref(2));
              label(1);
              fprintf(stream, "cmp $32,%%cl\n\t");
              fprintf(stream, "je "ref(3));
              // greater than 32-bit shift
              if (setflags && islogic)
              {
                fprintf(stream, "btr $29,(%%eax)\n\t"); // clear carry
              }
              fprintf(stream, "xorl %%edx,%%edx\n\t");  // clear edx
              fprintf(stream, "jmp "ref(2));
              label(3);
              if (setflags && islogic)
              {
                fprintf(stream, "bt $0,%%edx\n\t");
                fprintf(stream, "setc %%cl\n\t");
                fprintf(stream, "shll $29,%%ecx\n\t");
                fprintf(stream, "btr $29,(%%eax)\n\t");
                fprintf(stream, "orl %%ecx,(%%eax)\n\t");
              }
              fprintf(stream, "xorl %%edx,%%edx\n\t");
              label(2);
              break;
              
              case 2:  // asr
              fprintf(stream, "movzbl 12(%%esp),%%ecx\n\t");
              fprintf(stream, "movl 8(%%esp),%%edx\n\t");
              fprintf(stream, "cmp $31,%%cl\n\t");
              fprintf(stream, "jg "ref(1));
              fprintf(stream, "sarl %%cl,%%edx\n\t");
              if (setflags && islogic)
              {
                fprintf(stream, "setc %%cl\n\t");
                fprintf(stream, "shll $29,%%ecx\n\t");
                fprintf(stream, "movl 16(%%esp),%%eax\n\t");
                fprintf(stream, "btr $29,(%%eax)\n\t");
                fprintf(stream, "orl %%ecx,(%%eax)\n\t");
              }
              fprintf(stream, "jmp "ref(2));
              label(1);
              // greater than 31-bit shift
              fprintf(stream, "bt $31,%%edx\n\t");
              fprintf(stream, "xorl %%edx,%%edx\n\t");  // clear edx
              fprintf(stream, "sbbl $0,%%edx\n\t");
              if (setflags && islogic)
              {
                fprintf(stream, "movl %%edx,%%eax\n\t");
                fprintf(stream, "andl $0x20000000,%%eax\n\t");
                fprintf(stream, "btr $29,(%%eax)\n\t");
                fprintf(stream, "orl %%eax,(%%eax)\n\t");
              }
              label(2);
              break;
              
              case 3:  // ror
              fprintf(stream, "movzbl 12(%%esp),%%ecx\n\t");
              fprintf(stream, "movl 8(%%esp),%%edx\n\t");
              // register-valued ROR of zero doesn't seem to affect C
              if (setflags && islogic)
              {
                fprintf(stream, "cmp $0,%%cl\n\t");
                fprintf(stream, "je "ref(1));
              }
              fprintf(stream, "rorl %%cl,%%edx\n\t");
              if (setflags && islogic)
              {
                fprintf(stream, "setc %%cl\n\t");
                fprintf(stream, "shll $29,%%ecx\n\t");
                fprintf(stream, "movl 16(%%esp),%%eax\n\t");
                fprintf(stream, "btr $29,(%%eax)\n\t");
                fprintf(stream, "orl %%ecx,(%%eax)\n\t");
                label(1);
              }
              break;
            }
          }
          else  // immediate shift
          {
            switch (shifttype)
            {
              case 0:  // lsl
              fprintf(stream, "movl 12(%%esp),%%ecx\n\t");
              fprintf(stream, "movl 8(%%esp),%%edx\n\t");
              fprintf(stream, "shll %%cl,%%edx\n\t");
              if (setflags && islogic)
              {
                fprintf(stream, "setc %%cl\n\t");
                fprintf(stream, "shll $29,%%ecx\n\t");
                fprintf(stream, "movl 16(%%esp),%%eax\n\t");
                fprintf(stream, "btr $29,(%%eax)\n\t");
                fprintf(stream, "orl %%ecx,(%%eax)\n\t");
              }
              break;
              
              case 1:  // lsr
              if (setflags && islogic)
              {
                fprintf(stream, "movl 12(%%esp),%%ecx\n\t");
                fprintf(stream, "cmpl $0,%%ecx\n\t");
                fprintf(stream, "je "ref(1));
                fprintf(stream, "movl 8(%%esp),%%edx\n\t");
                fprintf(stream, "shrl %%cl,%%edx\n\t");
                fprintf(stream, "setc %%cl\n\t");
                fprintf(stream, "shll $29,%%ecx\n\t");
                fprintf(stream, "movl 16(%%esp),%%eax\n\t");
                fprintf(stream, "btr $29,(%%eax)\n\t");
                fprintf(stream, "orl %%ecx,(%%eax)\n\t");
                fprintf(stream, "jmp "ref(2));
                label(1);
                fprintf(stream, "bt $31,%%edx\n\t");
                fprintf(stream, "setc %%cl\n\t");
                fprintf(stream, "shll $29,%%ecx\n\t");
                fprintf(stream, "btr $29,(%%eax)\n\t");
                fprintf(stream, "orl %%ecx,(%%eax)\n\t");
                label(2);
              }
              else
              {
                fprintf(stream, "movl 12(%%esp),%%ecx\n\t");
                fprintf(stream, "cmpl $0,%%ecx\n\t");
                fprintf(stream, "je "ref(1));
                fprintf(stream, "movl 8(%%esp),%%edx\n\t");
                fprintf(stream, "shrl %%cl,%%edx\n\t");
                fprintf(stream, "jmp "ref(2));
                label(1);
                fprintf(stream, "movl 4(%%esp),%%eax\n\t");
                fprintf(stream, "ret\n");
                label(2);
              }
              break;
              
              case 2:  // asr
              if (setflags && islogic)
              {
                fprintf(stream, "movl 12(%%esp),%%ecx\n\t");
                fprintf(stream, "cmpl $0,%%ecx\n\t");
                fprintf(stream, "je "ref(1));
                fprintf(stream, "movl 8(%%esp),%%edx\n\t");
                fprintf(stream, "sarl %%cl,%%edx\n\t");
                fprintf(stream, "setc %%cl\n\t");
                fprintf(stream, "shll $29,%%ecx\n\t");
                fprintf(stream, "movl 16(%%esp),%%eax\n\t");
                fprintf(stream, "btr $29,(%%eax)\n\t");
                fprintf(stream, "orl %%ecx,(%%eax)\n\t");
                fprintf(stream, "jmp "ref(2));
                label(1);
                fprintf(stream, "bt $31,%%edx\n\t");
                fprintf(stream, "setc %%cl\n\t");
                fprintf(stream, "sarl $31,%%edx\n\t");
                fprintf(stream, "shll $29,%%ecx\n\t");
                fprintf(stream, "btr $29,(%%eax)\n\t");
                fprintf(stream, "orl %%ecx,(%%eax)\n\t");
                label(2);
              }
              else
              {
                fprintf(stream, "movl 12(%%esp),%%ecx\n\t");
                fprintf(stream, "cmpl $0,%%ecx\n\t");
                fprintf(stream, "je "ref(1));
                fprintf(stream, "movl 8(%%esp),%%edx\n\t");
                fprintf(stream, "sarl %%cl,%%edx\n\t");
                fprintf(stream, "jmp "ref(2));
                label(1);
                fprintf(stream, "movl 4(%%esp),%%eax\n\t");
                fprintf(stream, "ret\n");
                label(2);
              }
              break;
              
              case 3:  // ror
              fprintf(stream, "movl 12(%%esp),%%ecx\n\t");
              fprintf(stream, "movl 8(%%esp),%%edx\n\t");
              fprintf(stream, "movl 16(%%esp),%%eax\n\t");
              fprintf(stream, "cmpl $0,%%ecx\n\t");
              fprintf(stream, "je "ref(1));
              fprintf(stream, "rorl %%cl,%%edx\n\t");
              fprintf(stream, "jmp "ref(2));
              label(1);
              fprintf(stream, "bt $29,(%%eax)\n\t");
              fprintf(stream, "rcrl $1,%%edx\n");
              label(2);
              // carry valid for either ROR or RRX operation
              if (setflags && islogic)
              {
                fprintf(stream, "setc %%cl\n\t");
                fprintf(stream, "shll $29,%%ecx\n\t");
                fprintf(stream, "btr $29,(%%eax)\n\t");
                fprintf(stream, "orl %%ecx,(%%eax)\n\t");
              }
              break;
            }
          }

          fprintf(stream, "movl 4(%%esp),%%eax\n\t");

          switch (op)
          {
            case dp_AND:
            fprintf(stream, "andl %%edx,%%eax\n\t");
            break;
            
            case dp_EOR:
            fprintf(stream, "xorl %%edx,%%eax\n\t");
            break;
            
            case dp_SUB:
            fprintf(stream, "subl %%edx,%%eax\n\t");
            invc = 1;
            break;
            
            case dp_RSB:
            fprintf(stream, "xchgl %%edx,%%eax\n\t");
            fprintf(stream, "subl %%edx,%%eax\n\t");
            invc = 1;
            break;
            
            case dp_ADD:
            fprintf(stream, "addl %%edx,%%eax\n\t");
            break;
            
            case dp_ADC:
            fprintf(stream, "movl 16(%%esp),%%ecx\n\t");
            fprintf(stream, "bt $29,(%%ecx)\n\t");
            fprintf(stream, "adcl %%edx,%%eax\n\t");
            break;
            
            case dp_SBC:
            fprintf(stream, "movl 16(%%esp),%%ecx\n\t");
            fprintf(stream, "bt $29,(%%ecx)\n\t");
            fprintf(stream, "cmc\n\t");
            fprintf(stream, "sbbl %%edx,%%eax\n\t");
            invc = 1;
            break;
            
            case dp_RSC:
            fprintf(stream, "movl 16(%%esp),%%ecx\n\t");
            fprintf(stream, "xchgl %%edx,%%eax\n\t");
            fprintf(stream, "bt $29,(%%ecx)\n\t");
            fprintf(stream, "cmc\n\t");
            fprintf(stream, "sbbl %%edx,%%eax\n\t");
            invc = 1;
            break;
            
            case dp_TST:
            fprintf(stream, "testl %%edx,%%eax\n\t");
            break;
            
            case dp_TEQ:
            // leaving eax intact!
            fprintf(stream, "movl %%eax,%%ecx\n\t");
            fprintf(stream, "xorl %%edx,%%ecx\n\t");
            break;
            
            case dp_CMP:
            fprintf(stream, "cmpl %%edx,%%eax\n\t");
            invc = 1;
            break;
            
            case dp_CMN:
            fprintf(stream, "movl %%eax,%%ecx\n\t");
            fprintf(stream, "addl %%edx,%%ecx\n\t");
            break;
            
            case dp_ORR:
            fprintf(stream, "orl %%edx,%%eax\n\t");
            break;
            
            case dp_MOV:
            fprintf(stream, "movl %%edx,%%eax\n\t");
            needtest = 1;
            break;
            
            case dp_BIC:
            fprintf(stream, "notl %%edx\n\t");
            fprintf(stream, "andl %%edx,%%eax\n\t");
            break;
            
            case dp_MVN:
            fprintf(stream, "notl %%edx\n\t");
            fprintf(stream, "movl %%edx,%%eax\n\t");
            needtest = 1;
            break;
          }
          if (setflags)
          {
            if (islogic)
            {
              if (needtest)
              {
                fprintf(stream, "testl %%eax,%%eax\n\t");
              }
              fprintf(stream, "sete %%dl\n\t");
              fprintf(stream, "sets %%dh\n\t");
              fprintf(stream, "shl $1,%%dh\n\t");
              fprintf(stream, "or %%dh,%%dl\n\t");
              fprintf(stream, "shl $30,%%edx\n\t");
              fprintf(stream, "movl 16(%%esp),%%ecx\n\t");
              fprintf(stream, "andl $0x3fffffff,(%%ecx)\n\t");
              fprintf(stream, "orl %%edx,(%%ecx)\n\t");
            }
            else  // ALU instruction; all flags set
            {
              fprintf(stream, "seto %%dl\n\t");     // V
              if (invc)
                fprintf(stream, "setnc %%dh\n\t");  // !C
              else
                fprintf(stream, "setc %%dh\n\t");   // C
              fprintf(stream, "sete %%cl\n\t");     // Z
              fprintf(stream, "sets %%ch\n\t");     // N
              fprintf(stream, "shlb $1,%%dh\n\t");
              fprintf(stream, "shlb $2,%%cl\n\t");
              fprintf(stream, "shlb $3,%%ch\n\t");
              fprintf(stream, "orb %%dh,%%dl\n\t");
              fprintf(stream, "orb %%cl,%%dl\n\t");
              fprintf(stream, "orb %%ch,%%dl\n\t");
              fprintf(stream, "shll $28,%%edx\n\t");
              fprintf(stream, "movl 16(%%esp),%%ecx\n\t");
              fprintf(stream, "andl $0x0fffffff,(%%ecx)\n\t");
              fprintf(stream, "orl %%edx,(%%ecx)\n\t");
            }
          }
          fprintf(stream, "ret\n\n");
        }
      }
    }
  }
}

int main(void)
{
  gen_dp(stdout);

  return 0;
}
